### PR TITLE
Fix some mypy errors to get mypy working

### PR DESCRIPTION
**Pull request by oscarbenjamin**
#### Brief description of what is fixed or changed

This PR makes a lot of small changes and adds a lot of type-hints to make it possible to use myp on the sympy codebase.

Particular changes:
* Contains a fix for using py3 syntax even though the PY2 import test is still there (see .travis.yml).
* Remove all uses of `with_metaclass`. These were needed for PY2 compatibility but are not needed now.
* Replace all uses of list for `__slots__` with tuple. I thought this was necessary for mypy. I no longer think that it is but I still think it's a good change because `__slots__` doesn't need to be mutable.
* Add type hints in lots of places that create empty dicts/lists - this is comething that mypy complains about.
* Ensure that all definitions of a function are equivalent from a type perspective when types are defined/imported conditionally.
* Add type:ignore in lots of places particularly in multipledispatch.
* Tell mypy to ignore some internal code (rubi, antr, benchmarks and pygletplot) because they are autogenerated or unmaintained.

#### Other comments

This PR is the minimum needed to get mypy working so that we can *begin* adding type-hints and finding/preventing bugs.

At this stage I don't recommend making mypy a compulsory part of the Travis test suite because it is confusing. It could be added as an allowed failure but not in this PR which is intentionally minimal.

#### Release Notes

NO ENTRY

### CODE REVIEW

File: examples/intermediate/mplot3d.py [hunk 1]
```diff
@@ -24,7 +24,7 @@ def mplot3d(f, var1, var2, show=True):
     p = import_module('pylab')
     # Try newer version first
     p3 = import_module('mpl_toolkits.mplot3d',
-        __import__kwargs={'fromlist': ['something']}) or import_module('matplotlib.axes3d')
+        import_kwargs={'fromlist': ['something']}) or import_module('matplotlib.axes3d')
```
Comment (oscarbenjamin) [inline examples/intermediate/mplot3d.py]: In mypy an argument that begins with two underscores is a "positional-only" argument.

**Comment by asmeurer** [comment]
I'm interested if this catches any bugs.

**Comment by oscarbenjamin** [comment]
Apart from picking up on invalid imports in XFAIL tests the changes so far are mostly just style. It seems to complain about the same things that flake8 does but is more picky. This is only the first step though.

Without adding any type annotations `$ mypy sympy` gives 1000 errors on master. The changes so far here are just the minimum to eliminate half of those. Once this is done we could think about actually adding annotations for the types we care about.

When we get to adding type annotations I think that this will pick up some bugs. I expect the real benefits to come from new code rather than fixing existing code though. Many of the places in sympy that are inconsistent about types are also just poor API or hacky implementation. If we can avoid introducing more of that in future then I think new code will be better as a result.

**Comment by oscarbenjamin** [comment]
One thing that mypy picks up on is when related classes have different signatures for a method. For example it has found that `Matrix.evalf(prec=20)` is inconsistent with `Expr.evalf(n=20)`

**Comment by oscarbenjamin** [comment]
Another thing that mypy doesn't like is subclassing an alias of a class e.g.:
```python
class MutableDenseMatrix(...):
    pass

Matrix = MutableDenseMatrix

class SomeMatrix(Matrix):   # <---- this should be MutableDenseMatrix not Matrix
    passs
```

**Comment by oscarbenjamin** [comment]
Another thing that it doesn't like is the `V.i` syntax in vectors. The `Vector` constructor dynamically decides what kind of coordinate system it uses and then creates attributes like `x, y, z` or `r, theta, phi`. There are tests asserting that the attributes don't exist when a different coordinate system is used. I think that this is a poor design: there should be a class for each coordinate system.

**Comment by oscarbenjamin** [comment]
Another thing that mypy doesn't like is manipulation of `__doc__` e.g.
```
def f():
    "some stuff"

def g():
    pass

g.__doc__ = f.__doc__
```
Simple examples like above might pass fine but many similar examples in the sympy codebase lead to errors under mypy.

**Comment by oscarbenjamin** [comment]
I've edited the OP. This is ready for review.

This PR eliminates *all* roughly 1000 errors seen running mypy on master and should make it possible to experiment with mypy going forwards.

This changes many parts of the codebase so this PR will become stale and need merge conflicts resolving if not merged soon. Thanks for looking!

**Comment by jksuom** [comment]
These changes look good to me. They even seem to improve readability. I have also felt that instance attributes, also those that are not private like `__slots__`, should be tuples rather than lists. I think this could be merged.

**Comment by oscarbenjamin** [comment]
Thanks for reviewing @jksuom. I've been back through the diff myself and I'll merge it now.

Once this is in we can experiment with actually making proper use of type hints.

This also makes it possible to merge PRs with PY3 syntax.

**Pull request merged**